### PR TITLE
Feature/documentation

### PR DESCRIPTION
**Pull request by contested-space**
This adds a little bit of information to the README file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,104 @@
...
+
+## Usage
+-----
+
+For now, the only working mode is Carbon.
+
+need env :
+```
+  {erl_optics, [
+    {hostname, "<address of carbon host>"},
```
Comment (rkallos) [inline README.md]: I think it's better to have well-formatted yet obvious examples here, e.g `https://my-carbon-host.com`

File: README.md [hunk 2]
```diff
@@ -1,13 +1,104 @@
...
+    {port, <port number>},
+    {mode, carbon},
+    {interval, <polling interval time (ms)>}
+    ]}
```
Comment (rkallos) [inline README.md]: Indentation :trollface:

File: README.md [hunk 3]
```diff
@@ -1,13 +1,104 @@
...
+  Lenses = [erl_optics_lens:histo(<<"histo_name">>, [10, 20, 30, 40]),
+            erl_optics_lens:quantile(<<"quantile_name">>, 0.50, 10.0, 0.5)],
+  erl_optics:start_optics(<<"optics_namespace">>, Lenses).
+```
+
+## Lenses
+
+### counter
+A simple counter, polling will give the total increment value. Increment
+defaults to 1.
```
Comment (rkallos) [inline README.md]: Line wrapping? Not that I care much, since it's Markdown, but `M-x` `fill-region` or `fill-paragraph` are your friends :)

File: README.md [hunk 4]
```diff
@@ -1,13 +1,104 @@
...
+```
+  erl_optics:counter_inc(<<"<counter_name>">> [, <increment>]),
```
Comment (rkallos) [inline README.md]: I think it'd be better to have 2 different lines showing the two arities of that function. While using braces for optional stuff is succinct, I think it creates cognitive dissonance with Erlang syntax here.

File: README.md [hunk 5]
```diff
@@ -1,13 +1,104 @@
...
+### gauge
+Will keep the last value given to it.
+```
+  erl_optics:gauge_set(<<"<gauge_name>">>, <value>),
+```
+
+### dist
+Gives an estimation of the distribution of the values given to it. It provides
+metrics for 50th, 90th and 99th percentiles, the total number of values fed to
+it and the maximum value.
```
Comment (RAttab) [inline README.md]: I think the min is also provided

File: README.md [hunk 6]
```diff
@@ -1,13 +1,104 @@
...
+  % some things happen here
+  erl_optics:dist_record_timing_now_us(<<"dist_name">>, Time),
+```
+
+### histo
+Histo is a collection of buckets who are incremented every time a value
+within their boundaries is received. It is necessary to declare these
+lenses when starting optics:
+```
+  erl_optics_lens:histo(<<"<histo_name>">>, <buckets_list>)
```
Comment (RAttab) [inline README.md]: there are technically limitations around the bucket list and how to provide them. That being said, I wouldn't spend too much time on this lens as I'm not sure if it's useful at all.

File: README.md [hunk 7]
```diff
@@ -1,13 +1,104 @@
...
+```
+To add a value to the histogram:
+```
+  erl_optics:histo_inc(<<"<histo_name>">>, <value>),
+```
+
+### quantile
+Quantile is an estimation of the median value for a given percentile (e.g. if the
+percentile is 0.8, and the median is 10.0, that means that ~80% of values were
+below 10.0). It is necessary to declare these lenses when tarting optics.
```
Comment (rkallos) [inline README.md]: typo: tarting

File: README.md [hunk 8]
```diff
@@ -1,13 +1,104 @@
...
+```
+  erl_optics_lens:histo(<<"<histo_name>">>, <buckets_list>)
+```
+To add a value to the histogram:
+```
+  erl_optics:histo_inc(<<"<histo_name>">>, <value>),
+```
+
+### quantile
+Quantile is an estimation of the median value for a given percentile (e.g. if the
```
Comment (RAttab) [inline README.md]: I don't want to argue with a Math person but "the median value for a given percentile" sounds very wrong. In my brain the median is the 50th percentile so I'm unable to reasonably parse this sentence.

File: test/erl_optics_unit_tests.erl [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+start_stop_test() ->
+    start(),
+    check_in_foil(optics),
+    stop(),
+    ?assertEqual({error, module_not_found}, foil:lookup(erl_optics, optics)).
+
+counter_inc_test() ->
+    start(),
+    erl_optics:counter_inc(<<"counter">>),
+    foil_reload(),
```
Comment (rkallos) [inline test/erl_optics_unit_tests.erl:22]: Would it make sense to have an assert here for `<<"counter">>` not being in foil, as it would be in the ETS table instead?

File: README.md (hunk 8 above)
Comment (contested-space) [reply README.md]: Yeah you're right. I think the word I was looking for is "quantile"...

**Review by rkallos — COMMENTED** [review]
Thanks for kick-starting this! I have some suggestions about the docs, and I would also like to see one or more unit tests for quantiles.

File: README.md (hunk 7 above)
Comment (RAttab) [reply README.md]: oooh, what's tarting? Sounds delicious.

File: README.md (hunk 5 above)
Comment (contested-space) [reply README.md]: Not that I know of...
https://github.com/RAttab/optics/blob/master/src/optics.h#L118

Comment (RAttab) [reply README.md]: Oh wow you're right. Welp, I think I'm making you the new maintainer of optics as you clearly know it better then me. Congratulations.

File: test/erl_optics_unit_tests.erl (hunk 9 above)
Comment (contested-space) [reply test/erl_optics_unit_tests.erl:22]: I agree, but I think it warrants its own test, which I added under the foil_reload_test name.

File: README.md [hunk 10]
```diff
@@ -1,13 +1,137 @@
-erl_optics
-=====
+# erl_optics
 
-An OTP library
+An OTP application for [optics](https://github.com/RAttab/optics).
```
Comment (gnuvince) [inline README.md]: I'd add a few words to say what optics is.  E.g.:

> An OTP application for optics, a library to produce time series metrics.

File: README.md [hunk 11]
```diff
@@ -1,13 +1,137 @@
...
+## Building
 -----
-
     $ rebar3 compile
+
+## Usage
+-----
+
+For now, the only working mode is Carbon.
+
```
Comment (gnuvince) [inline README.md]: The mode is the back-end for which you can produce metrics, right?  If so, I'd rephrase to make it more clear:

> For now, erl-optics only supports the Carbon (link to Carbon) database.

File: README.md [hunk 12]
```diff
@@ -1,13 +1,137 @@
...
+need env :
```
Comment (gnuvince) [inline README.md]: Mention the name of the file, where it needs to be stored for erl-optics to pick it up.

File: README.md [hunk 13]
```diff
@@ -1,13 +1,137 @@
...
+```
+
+To start the optics polling:
+```
+  erl_optics:start_optics(<<"my_namespace">>),
+```
+or
+```
+  erl_optics:start_optics(<<"my_namespace">>, Lenses),
+```
```
Comment (gnuvince) [inline README.md:42]: What's the difference between these two invocations?

File: README.md [hunk 14]
```diff
@@ -1,13 +1,137 @@
...
+
+The namespace will be prepending the name of every metric, for
+example: `"optics_namespace.my_counter"`
```
Comment (gnuvince) [inline README.md]: In the code above you use `"my_namespace"`, so I'd rename this example to `"my_namespace.my_counter"`.

File: README.md [hunk 15]
```diff
@@ -1,13 +1,137 @@
...
+### counter
+A simple counter, polling will give the total increment value.
+Increment defaults to 1.
+```
+  erl_optics:counter_inc(<<"my_counter_name">>),
+```
+or
+```
+  erl_optics:counter_inc(<<"my_counter_name">, 10),
+```
```
Comment (gnuvince) [inline README.md:69]: Good.

File: README.md [hunk 16]
```diff
@@ -1,13 +1,137 @@
...
+
+### gauge
+Will keep the last value given to it.
```
Comment (gnuvince) [inline README.md]: I'm not sure I understand; if I don't produce a new value for a gauge, it's as if I keep sending the last value that I sent?

File: README.md [hunk 17]
```diff
@@ -1,13 +1,137 @@
...
+### dist
+Gives an estimation of the distribution of the values given to it. It
+provides metrics for 50th, 90th and 99th percentiles, the total number
+of values fed to it and the maximum value.
+```
+  erl_optics:dist_record(<<"my_dist_name">>, 10.0)
+```
+Dist can also be used to record timings with the two variants
+`dist_record_timing_now` (s) and `dist_record_timing_now_us` (micro seconds) that
+expect erlang timestamps as values and will record the difference
```
Comment (gnuvince) [inline README.md:86]: Write `(seconds)` and `(micro seconds)` or `(s)` and `(us)`.  Don't mix the two.

File: README.md [hunk 18]
```diff
@@ -1,13 +1,137 @@
...
+```
+  erl_optics_lens:quantile(<<"my_quantile_name">>, <target_quantile>, <estimated_value>, <adjustment_value>)
+```
+- target_quantile: float, wanted percentile.
+- estimated_value: expected value for given percentile
+- adjustment_value: will affect the speed at which the lens converges to the correct quantile value and the precision. (e.g. an adjustement of 0.1 means that the estimated quantile will rise by 0.1 when it receives a value over itself)
+
+## Polling
+In the carbon mode, polling will happen periodically at the time interval given in the parameters. When polling occurs,
+the metrics are sent at the carbon adress:port via TCP in a carbon-friendly format.
```
Comment (gnuvince) [inline README.md]: Typo: `address`.

File: README.md [hunk 19]
```diff
@@ -1,13 +1,137 @@
...
+
+If the lenses values need to be accessed directly, an erlang poller can be used instead
+```
+  erl_optics:register_erlang_poller(),
+```
```
Comment (gnuvince) [inline README.md:128]: I'm not sure what it means for the values to be accessed directly.

File: README.md (hunk 16 above)
Comment (contested-space) [reply README.md]: Exactly. Useful for things like states, where you could for example have a gauge showing 0 when a switch is off or 1 when it's on.

File: README.md (hunk 19 above)
Comment (contested-space) [reply README.md:128]: Instead of sending your metrics to carbon, this returns a map containing them.

File: README.md (hunk 12 above)
Comment (rkallos) [reply README.md]: This kinda depends on how the end user configures Erlang releases. In our case we'd put it in `env.config` in Salt, but other people do other things, perhaps even setting things manually in the invocation of `erl` via `-erl_optics foo bar`
